text/0000-dockerfiles.md

```diff
@@ -0,0 +1,113 @@
+# Meta
+[meta]: #meta
+- Name: Support Dockerfiles
+- Start Date: 2021-06-30
+- Author(s): sclevine
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: [RFC0069](https://github.com/buildpacks/rfcs/blob/main/text/0069-stack-buildpacks.md), [RFC#167](https://github.com/buildpacks/rfcs/pull/167)
+- Depends on: [RFC#172](https://github.com/buildpacks/rfcs/pull/172)
+
+# Summary
+[summary]: #summary
+
+This RFC introduces functionality for customizing base images, as an alternative to stackpacks.
+
+# Motivation
+[motivation]: #motivation
+
+Relying on Dockerfiles for base image generation and manipulation allows us to apply buildpacks only to the problem that they solve best: managing application runtimes and dependencies.
+
+# What it is
+[what-it-is]: #what-it-is
+
+This RFC proposes that we replace stackpacks with multi-purpose build-time and runtime Dockerfiles.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+Note: kaniko, BuildKit, and/or the original Docker daemon may be used to apply Dockerfiles at the platform's discretion. 
+
+### App-specified Dockerfiles
+
+A buildpack app may have a build.Dockerfile and/or run.Dockerfile in its app directory. A run.Dockerfile is applied to the selected runtime base image after the detection phase. A build.Dockerfile is applied to the build-time base image before the detection phase.
+
+Both Dockerfiles must accept `base_image` and `build_id` args. The `base_image` arg allows the lifecycle to specify the original base image. The `build_id` arg allows the app developer to bust the cache after a certain layer and must be defaulted to `0`.  
+
+A runtime base image may indicate that it preserves ABI compatibility by adding the label `io.buildpacks.rebasable=true`. In the case of app-specified Dockerfiles, `io.buildpacks.rebasable=false` is set automatically before `run.Dockerfile` is applied and must be explicitly set to `true` if desired. Rebasing an app without this label set to `true` requires passing a new `--force` flag to `pack rebase`.
+
+### Platform-specified Dockerfiles
+
+The same Dockerfiles may be used to create new stacks or modify existing stacks outside of the app build process. For both app-specified and stack-modifying Dockerfiles, any specified labels override existing values.
+
+Dockerfiles that are used to create a stack must create a `/cnb/stack/genpkgs` executable that outputs a [CycloneDX](https://cyclonedx.org)-formatted list of packages in the image with PURL IDs when invoked. This executable is executed after any run.Dockerfile or build.Dockerfile is applied, and the output replaces the label `io.buildpacks.sbom`. This label doubles as a Software Bill-of-Materials for the base image. In the future, this label will serve as a starting point for the application SBoM.
```

Comment:
why isn't `/cnb/stack/genpkgs` optional?

Reply:
If the "stack" concept is being removed in https://github.com/buildpacks/rfcs/pull/172 is this the right name for this dir?

Reply:
> why isn't /cnb/stack/genpkgs optional?

It's only required when creating a base image, not for app-specific Dockerfiles. I feel like it should be required so that changes from app-specific Dockerfiles (or from the base-image-creating Dockerfile) aren't silently dropped from the SBoM by accident. To opt-out, I'd like the base image author to explicitly creates `genpkgs` and make it no-op, so they understand that the SBoM will be inaccurate (which seems extremely dangerous, given that a CycloneDX SBoM strongly implies that the metadata can be used for vulnerability scanning).

> If the "stack" concept is being removed in #172 is this the right name for this dir?

Definitely not, good catch. I should also drop the word "stack" from this RFC.

Reply:
Moved to `/cnb/image/genpkgs`

Reply:
> author to explicitly creates `genpkgs` and

How can we create such `genpkgs` executable ? Where it it documented ? Example available ?

Reply:
It's an executable that run on the base image and outputs CycloneDX JSON to stdout. Happy to add more detail.

Reply:
> outputs CycloneDX JSON to stdout

Will this executable called during the build step (executed by lifecycle creator) and will output the JSON file containing the packages of the `build` but also the `run` image ? 

How will it be possible next to use such CycloneDX JSON file(s) to document the content of an image ?
Example
```
pack inspect buildpack{run|build}image_name
-->
...
packages:
   - package1
   - package2
   - package3
   - ....
```

Reply:
> Will this executable called during the build step (executed by lifecycle creator) and will output the JSON file containing the packages of the build but also the run image ?

It's not executed during the build phase (i.e., after all Dockerfiles have been applied), because the base image can't be modified by buildpacks. Buildpacks will be able to contribute their own entries to the SBoM via the buildpack BoM interface. See: https://github.com/buildpacks/rfcs/pull/166

Reply:
> How will it be possible next to use such CycloneDX JSON file(s) to document the content of an image ?

The CycloneDX project/ecosystem has rich tooling for this. It would be great if `pack inspect` could take advantage of that tooling to introspect all types of images (app images, base images, buildpack images).

Reply:
Marking as resolved given that we're leaving SBOM generation up to the platform (see https://github.com/buildpacks/spec/pull/298#issuecomment-1099543096).